lighthouse-core/report/v2/renderer/details-renderer.js

```diff
@@ -15,20 +15,85 @@
  */
 'use strict';
 
-/* globals self */
+/* globals self Util */
+
+/**
+ * Create render context for critical-request-chain tree display.
+ * @param {!DetailsRenderer.CRCNode} tree
+ * @return {!{tree: DetailsRenderer.CRCNode, startTime: number, transferSize: number}}
+ */
+function initTree(tree) {
+  const transferSize = 0;
+  let startTime = 0;
+  const rootNodes = Object.keys(tree);
+  if (rootNodes.length > 0) {
+    const node = /** @type {!DetailsRenderer.CRCNode} */ (tree[rootNodes[0]]);
+    startTime = node.request.startTime;
+  }
+
+  return {tree, startTime, transferSize};
+}
+
+/**
+ * Helper to create context for each critical-request-chain node based on its
+ * parent. Calculates if this node is the last child, whether it has any
+ * children itself and what the tree looks like all the way back up to the
+ * root, so the tree markers can be drawn correctly.
+ * @param {!(DetailsRenderer.CRCNode|{number: DetailsRenderer.CRCRequest})} parent
+ * @param {string} id
+ * @param {(Array<boolean>|undefined)} treeMarkers
+ * @param {(boolean|undefined)} parentIsLastChild
+ * @param {number} startTime
+ * @param {number} transferSize
+ * @return {!DetailsRenderer.CRCSegment}
+ */
+function createSegment(parent, id, treeMarkers, parentIsLastChild, startTime, transferSize) {
+  const node = /** @type {!DetailsRenderer.CRCNode} */ (parent[id]);
+  const siblings = Object.keys(parent);
+  const isLastChild = siblings.indexOf(id) === (siblings.length - 1);
+  const hasChildren = Object.keys(node.children).length > 0;
+
+  // Copy the tree markers so that we don't change by reference.
+  const newTreeMarkers = Array.isArray(treeMarkers) ? treeMarkers.slice(0) : [];
+
+  // Add on the new entry.
+  if (typeof parentIsLastChild !== 'undefined') {
+    newTreeMarkers.push(!parentIsLastChild);
+  }
+
+  return {
+    node,
+    isLastChild,
+    hasChildren,
+    startTime,
+    transferSize: transferSize + node.request.transferSize,
+    treeMarkers: newTreeMarkers
+  };
+}
+
 
 class DetailsRenderer {
   /**
    * @param {!DOM} dom
+   * @param {!Document|!Element} templateContext
    */
-  constructor(dom) {
+  constructor(dom, templateContext) {
```

Comment:
afaik categoryrenderer doesnt take this in the constructor. just has the `set` method.